File: CLASE V - Listas Void/listas_int/c/lista.c
#include <stdio.h>
#include <stdlib.h>

#include "lista.h"

struct Lista{

    NodoPtr primero;

};

NodoPtr getPrimero(ListaPtr lista);
void setPrimero(ListaPtr lista, NodoPtr nuevoPrimero);


ListaPtr crearLista(){

    ListaPtr lista = malloc(sizeof(struct Lista));

    lista->primero = NULL;

    return lista;
};


void insertarPrimero(ListaPtr lista, int dato){

    NodoPtr nuevoNodo = crearNodo(dato, lista->primero);

    lista->primero = nuevoNodo;


};


void insertarUltimo(ListaPtr lista, int dato){

    NodoPtr nuevoNodo = crearNodo(dato, NULL);

    NodoPtr actual = lista->primero;
    if(actual == NULL){
        lista->primero = nuevoNodo;
    }else{
        while(getSiguiente(actual)!=NULL){

            actual = getSiguiente(actual);
        }

       setSiguiente(actual, nuevoNodo);
        }

}; //tarea


void insertarPosicion(ListaPtr lista, int dato, int posicion); //tarea


int obtenerPrimero(ListaPtr lista); //tarea
int obtenerUltimo(ListaPtr lista); //tarea
int obtenerPosicion(ListaPtr lista, int posicion); //tarea

int eliminarPrimero(ListaPtr lista);
int eliminarUltimo(ListaPtr lista); //tarea
int eliminarposicion(ListaPtr lista, int posicion); //tarea

int obtenerTamanio(ListaPtr lista){

    int tam = 0 ;
    NodoPtr actual = lista->primero;
    while (actual != NULL){

            tam = tam + 1;
            actual = getSiguiente(actual);

    }

    return tam;
}; //tarea

void ordenarLista(ListaPtr lista){

    int permutacion;
    NodoPtr nodo;
    NodoPtr ultimo = NULL;

    if(obtenerTamanio(lista)< 2){
        //ya esta ordenada, no hago nada
    } else {

        do{
            permutacion = 0 ;
            nodo = lista->primero;

            while (getSiguiente(nodo) != ultimo){

                if(getDato(nodo) >   getDato(getSiguiente(nodo))){


                    int aux = getDato(nodo);
                    setDato(nodo, getDato(getSiguiente(nodo)));
                    setDato(getSiguiente(nodo), aux);
                    permutacion = 1;

                }
                nodo = getSiguiente(nodo);

            }
            ultimo = nodo; //opcional!!!

        }while (permutacion!=0);


    }


};//tarea
ListaPtr ordenarListaCopia(ListaPtr lista);//tarea
ListaPtr duplicarLista(ListaPtr lista);//tarea

int buscarElemento(ListaPtr lista, int datoBuscado);//tarea

void insertarEnOrden(ListaPtr lista, int dato);//tarea, si la lista está ordenada cada vez que
//se inserta se inserta en la posición que corresponda


void mostrarLista(ListaPtr lista){

    printf("\n<LISTA> TAM:%d \n", obtenerTamanio(lista));
    NodoPtr actual = lista->primero;

    while (actual!=NULL){

        mostrarNodo(actual);
        actual = getSiguiente(actual);
    }
    printf("\n\n");



};

void liberarLista(ListaPtr lista) {

    NodoPtr actual = lista->primero;
    NodoPtr aux;

    while (actual!=NULL){
        aux = actual;
        actual = getSiguiente(actual);
        free(aux);
    };
}

File: CLASE V - Listas Void/listas_int/golang/lista.go
package main

import "fmt"

type Lista struct {
	primero *Nodo
}

func crearLista() *Lista {
	return &Lista{primero: nil}
}

func insertarPrimero(lista *Lista, dato int) {
	lista.primero = crearNodo(dato, lista.primero)
}

func insertarUltimo(lista *Lista, dato int) {
	nuevo := crearNodo(dato, nil)
	actual := lista.primero
	if actual == nil {
		lista.primero = nuevo
		return
	}

	for getSiguiente(actual) != nil {
		actual = getSiguiente(actual)
	}
	setSiguiente(actual, nuevo)
}

func obtenerTamanio(lista *Lista) int {
	tam := 0
	actual := lista.primero
	for actual != nil {
		tam++
		actual = getSiguiente(actual)
	}
	return tam
}

func ordenarLista(lista *Lista) {
	if obtenerTamanio(lista) < 2 {
		return
	}

	var ultimo *Nodo
	for {
		permutacion := false
		nodo := lista.primero
		for getSiguiente(nodo) != ultimo {
			if getDato(nodo) > getDato(getSiguiente(nodo)) {
				aux := getDato(nodo)
				setDato(nodo, getDato(getSiguiente(nodo)))
				setDato(getSiguiente(nodo), aux)
				permutacion = true
			}
			nodo = getSiguiente(nodo)
		}
		ultimo = nodo
		if !permutacion {
			break
		}
	}
}

func mostrarLista(lista *Lista) {
	fmt.Printf("\n<LISTA> TAM:%d \n", obtenerTamanio(lista))
	actual := lista.primero
	for actual != nil {
		mostrarNodo(actual)
		actual = getSiguiente(actual)
	}
	fmt.Printf("\n\n")
}

